chia/wallet/wallet_state_manager.py

```diff
@@ -535,55 +554,190 @@ async def fetch_parent_and_check_for_cat(
             return None, None
         parent_coin_state = response[0]
         assert parent_coin_state.spent_height == coin_state.created_height
-        wallet_id = None
-        wallet_type = None
-        cs: Optional[CoinSpend] = await self.wallet_node.fetch_puzzle_solution(
+
+        coin_spend: Optional[CoinSpend] = await self.wallet_node.fetch_puzzle_solution(
             peer, parent_coin_state.spent_height, parent_coin_state.coin
         )
-        if cs is None:
+        if coin_spend is None:
             return None, None
-        matched, curried_args = match_cat_puzzle(Program.from_bytes(bytes(cs.puzzle_reveal)))
 
-        if matched:
-            mod_hash, tail_hash, inner_puzzle = curried_args
-            inner_puzzle_hash = inner_puzzle.get_tree_hash()
-            self.log.info(
-                f"parent: {parent_coin_state.coin.name()} inner_puzzle_hash for parent is {inner_puzzle_hash}"
-            )
+        # Check if the coin is a CAT
+        cat_matched, cat_curried_args = match_cat_puzzle(Program.from_bytes(bytes(coin_spend.puzzle_reveal)))
+        if cat_matched:
+            return await self.handle_cat(cat_curried_args, parent_coin_state, coin_state, coin_spend)
+
+        # Check if the coin is a NFT
+        #                                                        hint
+        # First spend where 1 mojo coin -> Singleton launcher -> NFT -> NFT
+        try:
+            uncurried_nft = UncurriedNFT.uncurry(Program.from_bytes(bytes(coin_spend.puzzle_reveal)))
+        except Exception:
+            # This is not a NFT coin, skip NFT handling
+            pass
+        else:
+            return await self.handle_nft(coin_spend, uncurried_nft)
 
-            hint_list = compute_coin_hints(cs)
-            derivation_record = None
-            for hint in hint_list:
-                derivation_record = await self.puzzle_store.get_derivation_record_for_puzzle_hash(bytes32(hint))
-                if derivation_record is not None:
-                    break
+        # Check if the coin is a DID
+        did_matched, did_curried_args = match_did_puzzle(Program.from_bytes(bytes(coin_spend.puzzle_reveal)))
+        if did_matched:
+            return await self.handle_did(did_curried_args, parent_coin_state, coin_state, coin_spend)
+
+        return None, None
+
+    async def handle_cat(
```

Comment:
Looks like we have significant changes here, so we should test to see if untrusted sync with large wallet (100k transactions) still works